api/adc/types.go

```diff
@@ -505,6 +505,20 @@ func (n *UpstreamNodes) UnmarshalJSON(p []byte) error {
 	return nil
 }
 
+// MarshalJSON implements the json.Marshaler interface for UpstreamNodes.
+// By default, Go serializes a nil slice as JSON null. However, for compatibility
+// with APISIX semantics, we want a nil UpstreamNodes to be encoded as an empty
+// array ([]) instead of null. Non-nil slices are marshaled as usual.
+//
+// See APISIX upstream nodes schema definition for details:
+// https://github.com/apache/apisix/blob/77dacda31277a31d6014b4970e36bae2a5c30907/apisix/schema_def.lua#L295-L338
+func (n UpstreamNodes) MarshalJSON() ([]byte, error) {
+	if n == nil {
+		return []byte("[]"), nil
```

Comment:
Isn't this achieved by filling empty slices as fallback values? This approach doesn't seem conducive to maintenance.
An overridden implementation would require digging deep into the code to discover.


https://go.dev/play/p/rB6BCY0n670

Reply:
* UpstreamNodes is part of the external API contract; the schema requires it to always be an array ([]), never null.
* Overriding MarshalJSON ensures consistent output regardless of whether the slice is initialized.
* This is defensive programming, preventing new code or omissions from causing the API to return null and break the contract.
* Compared to relying on every construction point to initialize the slice, this approach centralizes the handling and reduces maintenance overhead.
* The comment clearly explains the purpose and includes a schema link for reference.

https://go.dev/play/p/lS8wac0gRQZ

Reply:
Perhaps it's best to do this during both translation and encoding.

The implementation of the Marshaler interface is implicit; without examining the function implementations on UpstreamNodes, one cannot understand its specific operations. This requires familiarity with the codebase to explain why this data structure diverges from standard Go JSON encoding behavior.
I agree this should serve as a fallback, but only as a fallback. This doesn't mean we should overlook potential errors in the translation process.